icefall/diagnostics.py

```diff
@@ -0,0 +1,284 @@
+import torch
+from torch import Tensor
+from torch import nn
+import math
+import random
+from typing import Tuple, List
+
+
+class TensorDiagnosticOptions(object):
+    """
+    Options object for tensor diagnostics:
+
+     Args:
+        memory_limit: the maximum number of bytes per tensor (limits how many copies
+                of the tensor we cache).
```

Comment:
Please also update documentation about the arguments.

Reply:
Regarding print_pos_ratio: I think we can just hardcode this to true and remove the configuration.  So remove the stats_types() function and just use `["mean-abs", "pos-ratio"]`.

Reply:
> Regarding print_pos_ratio: I think we can just hardcode this to true and remove the configuration. So remove the stats_types() function and just use `["mean-abs", "pos-ratio"]`.
Could I modify the stats_types(self,) function to make it just return `["mean-abs", "pos-ratio"]`. I think it is safe since the stats types are fixed and could not be changed in this way. 

I am also confused about the condition `size != 31` defined in the dim_is_summarized(self, size:int) function.

Reply:
It's a heuristic about whether we want to print percentiles of the dimensions, for dimensions that don't have a "meaning" (e.g. relu, or attention heads), or whether we want to actually print the values one by one, for dimensions where the position has some kind of meaning, as in convolutional layers.

Reply:
> It's a heuristic about whether we want to print percentiles of the dimensions, for dimensions that don't have a "meaning" (e.g. relu, or attention heads), or whether we want to actually print the values one by one, for dimensions where the position has some kind of meaning, as in convolutional layers.

I see.. So the condition `size != 31` will be `False` and we'll print the convolutional kernel values one by one, since the   convolutional kernel size is set to 31 in Conformer.